tests/test_tables.py

```diff
@@ -0,0 +1,190 @@
+import wandb
+import pytest
+
+
+def test_basic_ndx():
+    # Base Case
+    table_a = wandb.Table(columns=["b"], data=[["a"], ["b"]])
+    table = wandb.Table(columns=["fi", "c"])
+    for _ndx, row in table_a.iterrows():
+        table.add_data(_ndx, "x")
+    assert all([row[0]._table == table_a for row in table.data])
+
+    # Adding is supported
+    table.add_data(3, "c")
+    # Adding duplicates is supported
+    table.add_data(3, "c")
+    # Adding None isn't supported
+    with pytest.raises(TypeError):
+        table.add_data(None, "d")
+
+    # Assert that the data in this column is valid, but also properly typed
+    assert [row[0] for row in table.data] == [0, 1, 3, 3]
+    assert all([row[0] is None or row[0]._table == table_a for row in table.data])
+
+
+def test_pk_cast(use_helper=False):
+    # Base Case
+    table = wandb.Table(columns=["id", "b"], data=[["1", "a"], ["2", "b"]])
+
+    # Validate that iterrows works as intended for no pks
+    assert [id_ for id_, row in list(table.iterrows())] == [0, 1]
+
+    # Cast as a PK
+    if use_helper:
+        table.set_pk("id")
+    else:
+        table.cast("id", wandb.data_types._TablePrimaryKeyType())
+
+    assert all(
+        [row[0]._table == table and row[0]._col_name == "id" for row in table.data]
```

Comment:
Ideally we wouldn't access underscore prefix methods like this, even in tests. (test the interface, not the implementation). Just a general comment, don't worry about it in this PR.

Reply:
That is fair, I should have made accessors on the special classes to get these without underscores. Noted for the future.